src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,370 @@
+    const iconSetConfigs = this._iconSetConfigs.get(namespace);
+    if (iconSetConfigs) {
+      return this._getSvgFromIconSetConfigs(name, this._iconSetConfigs.get(namespace));
+    }
+    return Observable.throw(new MdIconNameNotFoundException(iconKey));
+  }
+
+  /**
+   * Returns the cached icon for a SvgIconConfig if available, or fetches it from its URL if not.
+   */
+  private _getSvgFromConfig(config: SvgIconConfig): Observable<SVGElement> {
+    if (config.svgElement) {
+      // We already have the SVG element for this icon, return a copy.
+      return Observable.of(config.svgElement.cloneNode(true));
+    } else {
+      // Fetch the icon from the config's URL, cache it, and return a copy.
+      return this._loadSvgIconFromConfig(config)
+          .do((svg: SVGElement) => config.svgElement = svg)
+          .map((svg: SVGElement) => svg.cloneNode(true));
+    }
+  }
+
+  /**
+   * Attempts to find an icon with the specified name in any of the SVG icon sets.
+   * First searches the available cached icons for a nested element with a matching name, and
+   * if found copies the element to a new <svg> element. If not found, fetches all icon sets
+   * that have not been cached, and searches again after all fetches are completed.
+   * The returned Observable produces the SVG element if possible, and throws
+   * MdIconNameNotFoundException if no icon with the specified name can be found.
+   */
+  private _getSvgFromIconSetConfigs(name: string, iconSetConfigs: SvgIconConfig[]):
+      Observable<SVGElement> {
+    // For all the icon set SVG elements we've fetched, see if any contain an icon with the
+    // requested name.
+    const namedIcon = this._extractIconWithNameFromAnySet(name, iconSetConfigs);
+    if (namedIcon) {
+      // We could cache namedSvg in _svgIconConfigs, but since we have to make a copy every
+      // time anyway, there's probably not much advantage compared to just always extracting
+      // it from the icon set.
+      return Observable.of(namedIcon);
+    }
+    // Not found in any cached icon sets. If there are icon sets with URLs that we haven't
+    // fetched, fetch them now and look for iconName in the results.
+    const iconSetFetchRequests: Observable<SVGElement>[] = iconSetConfigs
+        .filter(iconSetConfig => !iconSetConfig.svgElement)
+        .map(iconSetConfig =>
+            this._loadSvgIconSetFromConfig(iconSetConfig)
+                .catch((err: any, source: any, caught: any): Observable<SVGElement> => {
+                  // Swallow errors fetching individual URLs so the combined Observable won't
+                  // necessarily fail.
+                  console.log(`Loading icon set URL: ${iconSetConfig.url} failed: ${err}`);
+                  return Observable.of(null);
+                })
+                .do((svg: SVGElement) => {
+                  // Cache SVG element.
+                  if (svg) {
+                    iconSetConfig.svgElement = svg;
```

Comment:
I reluctantly agree with @jelbourn. (On a related note, I'm very pleased with the 100-char line length).